channeldb/kvdb/etcd/commit_queue.go

```diff
@@ -131,20 +144,59 @@ func (c *commitQueue) Done(rset readSet, wset writeSet) {
 // dependencies. The queue ensures that the top element doesn't conflict with
 // any other transactions and therefore can be executed freely.
 func (c *commitQueue) mainLoop() {
-	defer c.wg.Done()
+	defer close(c.shutdown)
 
 	for {
+		// Wait until there are no unblocked transactions being
+		// executed, and for there to be at least one blocked
+		// transaction in our queue.
+		c.queueCond.L.Lock()
+		for c.freeCount > 0 || c.queue.Front() == nil {
+			c.queueCond.Wait()
```

Comment:
I did a few runs of `make itest etcd=1 icase=async_payments_benchmark` to see if there's any performance difference between this and the `RWMutex` based queue implementation and this seems to deadlock. I've uploaded the profile: https://paste.ubuntu.com/p/J3DZCDwCCQ/

To me it seems like that the problem might be that even though the context is canceled, we never signal the `queueCond`. We only signal when waiting for the queue to "stop".

Reply:
good catch @bhandras! I was able to figure it out, see the second commit :)